torchtune/utils/pooling.py

```diff
@@ -0,0 +1,49 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import Tensor
+
+
+def pool_sequence_logits(
+    tokens: Tensor, logits: Tensor, padding_token_idx: int = None
+) -> Tensor:
+    """Pool sequence logits by selecting the predicted logits for the last non-padding token
+    for each sequence in the batch.
+
+    If padding_token_id is not provided, the function assumes that the input tensor has a batch size of 1.
+    Args:
+        tokens (Tensor): input tensor with shape [b x s]
+        logits (Tensor): predicted logits for input tokens with shape [b x s x n]
+        padding_token_idx (int): Padding token id used in the tokenizer.
+    Returns:
+        Tensor: Pooled logits with shape [b x n]
+    Notation used for tensor shapes:
+        - b: batch size
+        - s: sequence length
+        - n: number of classes
+    Raises:
+        ValueError: If padding_token_idx is not provided and batch size is greater than 1.
+    """
+    batch_size = tokens.shape[0]
+
+    # padding token needs to be defined to pool multiple sequences
+    if padding_token_idx is None and batch_size != 1:
+        raise ValueError("padding_token_idx must be set if batch_size > 1")
+    if padding_token_idx is None:
+        sequence_lengths = -1
```

Comment:
This part I don't fully understand. Is the assumption here that for a batch size of 1 there will be no padding? This may hold in the current state of the world since we pad to the max length in the batch, but for general padding schemes (e.g. pad to fixed length) this may not hold. Personally I would just err on the side of caution and make padding_token_idx a required arg, then you don't have to branch based on batch_size (unless there's an important use case you're solving for here).

Reply:
The assumption is the other way round - if there's no padding, there _must be a batch size of 1_. Otherwise, you wouldn't have a way to grab the last token from multiple sequences in a batch with variable lengths.

If you do have a batch size of 1, the function is agnostic to this, it'll just give you the last token if you don't pass in a padding token, but it'll give you the last non-padding token otherwsie (the default behaviour). This is [how HF implements it](https://github.com/huggingface/transformers/blob/928331381ef6ce0622c0b1ac704299046b3afa21/src/transformers/models/mistral/modeling_mistral.py#L1339), I'll attribute it in the function for clarity. My intuition here is that the use case is for inference, potentially using a tokenizer where the padding token isn't defined, or you just generally know you have a full sequence/remove padding on your own side. 

Removing the branch for batch size wouldn't actually change how the sequence(s) are pooled, but just prevents the failure case. Allowing the function to be used without a padding token just gives a little flexibility.

Let me know if this makes sense. Open to your thoughts :)

Reply:
OK this generally makes sense. My main question is whether we will actually ever hit this function during inference; if its purpose is just to extract the EOS from an existing sequence it seems like we shouldn't, right? And yes attribution to the HF implementation would be great here

Reply:
Yeah, I thought about this for a bit and I think I'm with you on making it required. I think having to handle this error wherever its used is more annoying than just supporting a very nice usecase. 

> if its purpose is just to extract the EOS from an existing sequence

My usecase for performing inference was: run your trained `_classifier` model to obtain `[b, s, n]`, then having to call `pool_sequence_logits` to get `[b, n]`.